File: Markov/Link.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MarkovGenerator
{
    public class Link
    {
        private Random randy = new Random();

        public List<string> Before { get; set; } = new List<string>();
        public List<string> After { get; set; } = new List<string>();
        public string Word { get; set; }
        public int Frequency { get; set; }

        public Link()
        {
            Word = String.Empty;
            Frequency = 0;
        }

        public Link(string _word) // Beginning or end of file
        {
            Word = _word;
            Frequency = 0;
        }
        public Link(string _word, string _before, string _after)
        {
            Word = _word;
            Before.Add(_before);
            After.Add(_after);
            Frequency = 0;
        }

        public void AddBefore(string s)
        {
            Before.Add(s);
        }

        public void AddBefore(IEnumerable<string> lst)
        {
            foreach (string s in lst)
            {
                Before.Add(s);
            }
        }

        public void AddAfter(string s)
        {
            After.Add(s);
        }

        public void AddAfter(IEnumerable<string> lst)
        {
            foreach (string s in lst)
            {
                After.Add(s);
            }
        }

        public String RandomBefore()
        {
            try
            {
                var i = randy.Next(0, After.Count);
                return After[i];
            }
            catch // Empty list
            {
                return "~empty";
            }
        }

        public String RandomAfter()
        {
            try
            {
                var i = randy.Next(0, After.Count);
                return After[i];
            }
            catch
            {
                return "~empty";
            }
        }
    }
}

File: MarkovNextGen/Link.fs
namespace MarkovNextGen

open System
open System.Collections.Generic
open Newtonsoft.Json
open System.Linq

type public Link(lst : string list) =
    class
        let randy = new Random()
        let mutable after = lst

        new (lst : IEnumerable<string>) =
            Link(lst |> List.ofSeq)
    
        new () =
            Link([])
    
        member public this.After
            with get() = after
            and set (value) = after <- value
    
        // Way more efficient than adding each element individually
        member public this.AddAfter(lst : IEnumerable<string>) =
            let _lst = lst |> List.ofSeq
            after <- (List.append after _lst)
        
        member public this.AddAfter(lst : string list) =
            after <- List.append after lst
        
        member public this.AddAfter(s) =
            after <- s :: after
        
        [<JsonIgnore>]
        member public this.RandomAfter
            with get () =
                if after.Length > 0 then
                    let i = randy.Next(0, after.Length)
                    after.[i]
                else String.Empty
    end